sdk/provisioning/Azure.Provisioning/src/Expressions/AST.cs

```diff
@@ -0,0 +1,351 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Azure.Provisioning.Expressions;
+
+// This is a partial, simplified version of the Bicep AST that you can
+// reference at https://github.com/Azure/bicep/blob/main/docs/grammar.md.  It's
+// a non-goal to be able to do EVERYTHING that Bicep allows in C#.  We're just
+// trying to glue them together better - if you need something not supported
+// here then you can drop down to Bicep directly.
+
+public abstract class Expression
+{
+    internal abstract BicepWriter Write(BicepWriter writer);
+    public override string ToString() => new BicepWriter().Append(this).ToString();
+
+    public static implicit operator Expression(string value) => new StringLiteral(value);
+    public static implicit operator Expression(int value) => new IntLiteral(value);
+    public static implicit operator Expression(bool value) => new BoolLiteral(value);
+}
+
+public class IdentifierExpression(string name) : Expression
+{
+    public string Name { get; } = name;
+    internal override BicepWriter Write(BicepWriter writer) => writer.Append(Name);
+}
+
+public abstract class Literal(object? literalValue = null) : Expression
+{
+    public object? LiteralValue { get; } = literalValue;
+}
```

Comment:
Tucked away in its own namespace we've got a subset of the Bicep grammar that we can make use of.  Like most language grammars it's split into `Expression`s and `Statement`s.  `Expression` can go anywhere.  `Statement` expects to be top-level.  We're only implementing a subset of the language here for a couple of reasons.  First, if you really want everything Bicep offers then I'd happily direct to Bicep which solves all those problems for you today.  Second, the language is a moving target and I don't have a ton of confidence in our ability to stay on top of things in perpetuity.  Third, there's a ton of overlap between advanced Bicep features and C#.  I think most people would rather loop in C# than build a nested tree forming a Bicep `LoopStatement`.  I mostly started with what I needed and plan to let everything else grow organically from there.

Also, we can talk about the name being so general, but I'm fine with anyone wandering into this namespace being advanced enough to alias the whole namespace to `B` or something if they were mixing this with LINQ or an open source CAS or ....?